### PR TITLE
feat: harvest PR title from merge commit messages

### LINKED ISSUE #308: Get the NEWS item from the PR title
if not specified in the merge/squash commit message.

### PR DESCRIPTION
**Pull request by maelle**
Fix #308 

To replace #334

Still WIP

TODOS
- [x] Check there is a GitHub token. There is a similar check in auto.R
- [x] Add tests (with mocking)
- [x] Remove the demo edits of NEWS.md
- [ ] See https://github.com/cynkra/fledge/pull/334#discussion_r877596135
- [x] Regroup things into headers including "Uncategorized"

### CODE REVIEW

**Comment by krlmlr** [comment]
- [ ] Extract the PR number from the commit message
- [ ] Use it as part of the NEWS entry (also with placeholder entry in offline mode)

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,71 @@
...
+## Features
+
+Improve parsing of conventional commit messages
+
+
+## Bug fixes
+
+Fix conventional commit regex (#329)
+
+
```
Comment (maelle) [inline NEWS.md:13]: too many whitelines, I need to fix this

File: R/update-news.R [hunk 2]
```diff
@@ -45,6 +45,18 @@ remove_housekeeping <- function(message) {
 }
 
 extract_newsworthy_items <- function(message) {
+  if (is_merge_commit(message)) {
+    pr_data <- harvest_pr_data(message)
+    pr_number <- pr_data$pr_number
+    title <- pr_data$title %||% sprintf("https://github.com/%s/pull/%s", github_slug(), pr_number)
```
Comment (maelle) [inline R/update-news.R]: not replaceable yet but at least one can click on the link

**Comment by maelle** [comment]
Need to build a tibble with
- message
- header %||% uncategorized
- breaking or not

File: R/update-news.R [hunk 3]
```diff
@@ -171,3 +183,40 @@ conventional_commit_types <- function() {
...
+  }
+
+  tibble::tibble(
+    title = pr_info$title,
+    pr_number = pr_number
+  )
+}
+
+has_internet <- function() {
+  if (nzchar(Sys.getenv("NO_INTERNET_TEST_FLEDGE"))) {
```
Comment (maelle) [inline R/update-news.R:218]: this is a switch for tests

File: DESCRIPTION [hunk 4]
```diff
@@ -51,6 +52,7 @@ Suggests:
     digest,
     fansi,
     fs,
+    httptest,
```
Comment (maelle) [inline DESCRIPTION:55]: I used this and environmental variables for tests because I saw too many problems with mockery's depth argument.

File: NEWS.md [hunk 5]
```diff
@@ -1,5 +1,63 @@
 <!-- NEWS.md is maintained by https://cynkra.github.io/fledge, do not edit -->
```
Comment (maelle) [inline NEWS.md:1]: edits will be removed before merging

File: NEWS.md [hunk 6]
```diff
@@ -1,5 +1,63 @@
...
+
+- Improve parsing of conventional commit messages
+
+- improve bump_version() (error) messages (#328)
+
+Co-authored-by: maelle <maelle@users.noreply.github.com>
+
+- Improve bump_version() behavior in the absence of changes (#323)
+
+Co-authored-by: Kirill Müller <krlmlr@users.noreply.github.com>
```
Comment (maelle) [inline NEWS.md:19]: this could be used for #344

File: NEWS.md [hunk 7]
```diff
@@ -1,5 +1,63 @@
...
+
+- Turn Netlify builds off for now
+
+- Enable auto-style on GitHub Actions
+
+- Remove testthat specialization for snapshots
+
+
+
+## Uncategorized
```
Comment (maelle) [inline NEWS.md:38]: the grouping that does not work yet is the grouping of _scopes_

File: R/update-news.R [hunk 8]
```diff
@@ -19,39 +20,78 @@ collect_news <- function(messages) {
...
     purrr::map(extract_newsworthy_items) %>%
-    unlist()
+    purrr::keep(~ nrow(.) > 0) %>%
+    bind_rows()
 
-  if (length(newsworthy_items) == 0) {
+  if (is.null(newsworthy_items)) {
     if (length(messages) <= 1) {
-      newsworthy_items <- "- Same as previous version."
+      newsworthy_items <- parse_regular_commit("Same as previous version.")
```
Comment (maelle) [inline R/update-news.R:28]: the `parse_blabla_commits()` functions return tibbles with description, type, breaking, scope.

File: R/update-news.R [hunk 9]
```diff
@@ -19,39 +20,78 @@ collect_news <- function(messages) {
...
   strsplit(message, "\n---", fixed = TRUE)[[1]][1]
 }
 
 extract_newsworthy_items <- function(message) {
+  # Merge messages
+  if (is_merge_commit(message)) {
+    pr_data <- harvest_pr_data(message)
+    pr_number <- pr_data$pr_number
+    title <- if (is.na(pr_data$title)) {
+      sprintf("PLACEHOLDER https://github.com/%s/pull/%s", github_slug(), pr_number)
```
Comment (maelle) [inline R/update-news.R:55]: I added a link because this way one could click on it.
Should I already add a function for replacing these items i.e. retrying to harvest PR data?

File: tests/testthat/_snaps/update-news/NEWS.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+
+
+## Features 
+
+### lang 
+
+- add Polish language
+
+### api
```
Comment (maelle) [inline tests/testthat/_snaps/update-news/NEWS.md:24]: need to add uncategorized for scopes

File: DESCRIPTION [hunk 11]
```diff
@@ -28,6 +28,7 @@ Imports:
     callr,
     cli,
     clipr,
+    curl,
```
Comment (maelle) [inline DESCRIPTION:31]: With this PR, there are too many imports :grimacing:

File: R/update-news.R (hunk 9 above)
Comment (krlmlr) [reply R/update-news.R:55]: The clickable link is a good idea. Let's postpone extensions for now.

File: R/update-news.R [hunk 12]
```diff
@@ -19,39 +20,78 @@ collect_news <- function(messages) {
...
+  newsworthy_items
 }
 
 remove_housekeeping <- function(message) {
   strsplit(message, "\n---", fixed = TRUE)[[1]][1]
 }
 
 extract_newsworthy_items <- function(message) {
+  # Merge messages
+  if (is_merge_commit(message)) {
```
Comment (krlmlr) [inline R/update-news.R:51]: Should we introduce `parse_merge_commit()` and `parse_bullet_commit()`? The goal would be to have all code in this function "on the same level of abstraction".

File: R/update-news.R [hunk 13]
```diff
@@ -171,3 +207,122 @@ conventional_commit_types <- function() {
...
+      error = function(e) {
+        print(e)
+        cli::cli_alert_warning(failure_message)
+        return(NULL)
+      }
+    )
+  }
+
+  tibble::tibble(
+    title = pr_info$title %||% NA,
```
Comment (krlmlr) [inline R/update-news.R:242]: Minor nit -- for type stability:

```suggestion
    title = pr_info$title %||% NA_character_,
```

File: tests/testthat/_snaps/demo/demo.md [hunk 14]
```diff
@@ -315,6 +315,7 @@ cat(news, sep = "\n")
 ## # tea 0.0.0.9001
 ## 
 ## - Add tests for cup of tea.
+##
```
Comment (krlmlr) [inline tests/testthat/_snaps/demo/demo.md:318]: I find NEWS more concise if there is no empty line inbetween. Are there semantic differences?

File: tests/testthat/setup.R [hunk 15]
```diff
@@ -0,0 +1 @@
+library(httptest)
```
Comment (krlmlr) [inline tests/testthat/setup.R:1]: Why is this necessary? Can you please add a comment?

**Review by krlmlr — APPROVED** [review]
Thanks, looks good! Can you please take another quick look, and merge?
